manager_dashboard/manager_dashboard/js/usergroup-management.js

```diff
@@ -0,0 +1,150 @@
+// Clear input values when modal is to be shown
+$("#add-usergroup-modal").on('show.bs.modal', function() {
+  $('#usergroup-name').val('');
+  $('#usergroup-description').val('');
+});
+
+// Shows loading message in the modal and disables the inputs
+function showUserGroupLoading() {
+  $('#add-usergroup-modal-loading-message-container').removeClass('hidden');
+  $('#usergroup-name').attr('disabled', true);
+  $('#usergroup-description').attr('disabled', true);
+}
+
+// Hides loading message in the modal and enables the inputs
+function hideUserGroupLoading() {
+  $('#usergroup-name').attr('disabled', false);
+  $('#usergroup-description').attr('disabled', false);
+  $('#add-usergroup-modal-loading-message-container').addClass('hidden');
+}
+
+// Encodes the given string to the firebase key compatible string
+function encodeStringToKey(string) {
+  return encodeURIComponent(string).replace(/\./g, '%2E');
+}
+
+// Gets triggered when user hits submit button in add usergroup modal
+function submitNewUserGroup() {
+  const nameInput = $('#usergroup-name');
+  const name = (nameInput.val() || '').trim();
+  const descriptionInput = $('#usergroup-description');
+  const description = (descriptionInput.val() || '').trim();
+
+  if (name && name.length > 30) {
+    hideUserGroupLoading();
+    alert('User group name is required and should be less than 30 characters');
+  } else if (description && description.length > 100) {
+    hideUserGroupLoading();
+    alert('User group description is should be less than 100 characters');
+  } else {
+    showUserGroupLoading();
+
+    var key;
+    try {
+      key = encodeStringToKey(name);
+    } catch {
+      // This should never occur
+      alert('Failed to encode name into a key, please make sure that the user group name does not contain any invalid characters.');
+    }
+
+    if (key) {
+      firebase.database().ref('v2/userGroups/' + key).once("value", function(snapshot) {
+        if (snapshot.exists()) {
+          hideUserGroupLoading();
+          alert('A group with this name already exists');
+        } else {
+          var updates = {};
+          updates['/v2/userGroups/' + key] = {
```

Comment:
Thinking again about my comment during the call, I think using the user provided name is very likely to cause a lot of sadness down the line :sob: If a group is renamed (very likely, given how many users ask us to rename their user account), you would need to atomically change the key in multiple places within firebase. Not impossible, but cumbersome.

The other problem is that in the app, you get the list of groups as a `.once()` operation, without subscribing to updates. So if I open the list of groups, an admin renames the group I'm about to join, and then I join the group with the old name (because the app won't know about the rename unless I close and reopen the screen), you have a bug :bug: 

I think you could structure the data so that each group has a unique ID (generated in code, ideally a non sequential value so that enumeration is harder), and you should be able to check if the name already exists with something like `ref.child('userGroups').orderByChild('name').equalTo('MyNewGroup')`. What do you think? 
(this won't prevent simultaneous inserts of a single value if 2 admins create a group with the same name at the exact same time, but given that we have a handful of admins but thousands of users, it's a lot less likely to happen)
Firebase database does not seem to have anything like unique constraints, so I guess this is the best we can do.

Reply:
@laurentS the concerns you have completely makes sense, let discuss it further on that part. I've thought about the query that you mentioned.
```
ref.child('userGroups').orderByChild('name').equalTo('MyNewGroup')
```
The major problem with this one would be that we cannot make the case insensitive check. So 'MyUserGroup' would be treated differently from 'Myusergroup'.